### PR TITLE
tests(openshift/nightly): Change secret creation step in demo

### PR DESCRIPTION
**Pull request by nshankar13**
Signed-off-by: nshankar13 <nshankar@microsoft.com>

**Description**: Edit demo for openshift nightly job and add cleanup step. 

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [X] |
| CI System                  | [X] |
| Demo                       | [X] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? No
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change? No

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -170,9 +170,9 @@ export BOOKWAREHOUSE_NAMESPACE=bookwarehouse
 # Default: 90s
 # export TIMEOUT=90s
 
-### optional: Whether to publish OSM docker images to container registry
```
Comment (ksubrmnn) [inline .env.example:173]: You should still keep the PUBLISH_IMAGES flag. It's separate from whether or not a private registry is being used

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

File: .github/workflows/openshift-nightly.yml [hunk 2]
```diff
@@ -45,8 +45,19 @@ jobs:
           CI_MAX_WAIT_FOR_POD_TIME_SECONDS: 150
           CI_MIN_SUCCESS_THRESHOLD: 1
           OSM_HUMAN_DEBUG_LOG: true
-          PUBLISH_IMAGES: "false"
+          USE_PRIVATE_REGISTRY: "true"
```
Comment (ksubrmnn) [inline .github/workflows/openshift-nightly.yml:48]: the openservicemesh registry is pubish, right? So this should be false, because we don't need the secret for this so we also don't need to link it for pulling images

File: .github/workflows/openshift-nightly.yml [hunk 3]
```diff
@@ -45,8 +45,19 @@ jobs:
...
         run: |
           touch .env
           ./demo/run-osm-demo.sh
           go run ./ci/cmd/maestro.go
+      - name: Cleanup resources
```
Comment (ksubrmnn) [inline .github/workflows/openshift-nightly.yml:53]: can you move this out into a separate PR? this addresses a different issue

File: .env.example (hunk 1 above)
Comment (nshankar13) [reply .env.example:173]: I was considering that but we don't use it anywhere in the demo - should I still keep it?

File: demo/run-osm-demo.sh [hunk 4]
```diff
@@ -84,7 +84,7 @@ if [ "$DEPLOY_ON_OPENSHIFT" = true ] ; then
     optionalInstallArgs+=" --set=OpenServiceMesh.enablePrivilegedInitContainer=true"
 fi
 
-if [ "$PUBLISH_IMAGES" = true ]; then
+if [ "$USE_PRIVATE_REGISTRY" = false ]; then
     make docker-push
```
Comment (ksubrmnn) [inline demo/run-osm-demo.sh:88]: this should be wrapped around the PUBLISH_IMAGES condition

File: demo/run-osm-demo.sh [hunk 5]
```diff
@@ -84,7 +84,7 @@ if [ "$DEPLOY_ON_OPENSHIFT" = true ] ; then
...
     ./scripts/create-container-registry-creds.sh "$K8S_NAMESPACE"
```
Comment (ksubrmnn) [inline demo/run-osm-demo.sh:89]: This should be wrapped by the PRIVATE_REGISTRY condition

**Comment by ksubrmnn** [comment]
You should also modify the `create-container-registry-creds.sh` script to use the USE_PRIVATE_REGISTRY variable to determine whether or not to create the secret.

The script is currently doing that in a roundabout way by checking to see if the password is empty: https://github.com/openservicemesh/osm/blob/main/scripts/create-container-registry-creds.sh#L16

**Comment by nshankar13** [comment]
> You should also modify the create-container-registry-creds.sh script to use the USE_PRIVATE_REGISTRY variable to determine whether or not to create the secret.

@ksubrmnn I think it would make sense to do this in a different PR because multiple files use the create-container-registry-creds.sh file, and we would need to create the USE_PRIVATE_REGISTRY variables in these scripts as well.

File: demo/run-osm-demo.sh [hunk 6]
```diff
@@ -138,6 +142,13 @@ fi
 
 ./demo/configure-app-namespaces.sh
 
+if [[ "$USE_PRIVATE_REGISTRY" = true && "$DEPLOY_ON_OPENSHIFT" = true ]]; then
+    oc secrets link bookbuyer "$CTR_REGISTRY_CREDS_NAME" --for=pull -n "$BOOKBUYER_NAMESPACE"
```
Comment (ksubrmnn) [inline demo/run-osm-demo.sh:146]: I don't think this will work because the service accounts aren't created yet. You should instead wrap the existing calls to `oc secrets link` since those are in the correct place (called after serviceaccount creation)

**Comment by ksubrmnn** [comment]
> > You should also modify the create-container-registry-creds.sh script to use the USE_PRIVATE_REGISTRY variable to determine whether or not to create the secret.
> 
> @ksubrmnn I think it would make sense to do this in a different PR because multiple files use the create-container-registry-creds.sh file, and we would need to create the USE_PRIVATE_REGISTRY variables in these scripts as well.

Sure whatever is easiest for you.

File: demo/run-osm-demo.sh (hunk 6 above)
Comment (nshankar13) [reply demo/run-osm-demo.sh:146]: I see, I didn't realize it was already in the deploy-bookbuyer / store scripts

File: demo/deploy-bookbuyer.sh [hunk 7]
```diff
@@ -10,6 +10,7 @@ CI_CLIENT_CONCURRENT_CONNECTIONS="${CI_CLIENT_CONCURRENT_CONNECTIONS:-1}"
 ENABLE_EGRESS="${ENABLE_EGRESS:-false}"
 CI_SLEEP_BETWEEN_REQUESTS_SECONDS="${CI_SLEEP_BETWEEN_REQUESTS_SECONDS:-1}"
 DEPLOY_ON_OPENSHIFT="${DEPLOY_ON_OPENSHIFT:-false}"
+USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-true}"
```
Comment (ksubrmnn) [inline demo/deploy-bookbuyer.sh:13]:
```suggestion
USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-false}"
```

File: demo/deploy-bookstore.sh [hunk 8]
```diff
@@ -7,6 +7,7 @@ source .env
 VERSION=${1:-v1}
 SVC="bookstore-$VERSION"
 DEPLOY_ON_OPENSHIFT="${DEPLOY_ON_OPENSHIFT:-false}"
+USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-true}"
```
Comment (ksubrmnn) [inline demo/deploy-bookstore.sh:10]:
```suggestion
USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-false}"
```

File: demo/deploy-bookthief.sh [hunk 9]
```diff
@@ -12,6 +12,7 @@ CI_CLIENT_CONCURRENT_CONNECTIONS="${CI_CLIENT_CONCURRENT_CONNECTIONS:-1}"
 ENABLE_EGRESS="${ENABLE_EGRESS:-false}"
 CI_SLEEP_BETWEEN_REQUESTS_SECONDS="${CI_SLEEP_BETWEEN_REQUESTS_SECONDS:-1}"
 DEPLOY_ON_OPENSHIFT="${DEPLOY_ON_OPENSHIFT:-false}"
+USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-true}"
```
Comment (ksubrmnn) [inline demo/deploy-bookthief.sh:15]:
```suggestion
USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-false}"
```

File: demo/deploy-bookwarehouse.sh [hunk 10]
```diff
@@ -5,6 +5,8 @@ set -aueo pipefail
 # shellcheck disable=SC1091
 source .env
 DEPLOY_ON_OPENSHIFT="${DEPLOY_ON_OPENSHIFT:-false}"
+USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-true}"
```
Comment (ksubrmnn) [inline demo/deploy-bookwarehouse.sh:8]:
```suggestion
USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-false}"
```

File: demo/run-osm-demo.sh [hunk 11]
```diff
@@ -37,6 +37,7 @@ DEPLOY_WITH_SAME_SA="${DEPLOY_WITH_SAME_SA:-false}"
 ENVOY_LOG_LEVEL="${ENVOY_LOG_LEVEL:-debug}"
 DEPLOY_ON_OPENSHIFT="${DEPLOY_ON_OPENSHIFT:-false}"
 TIMEOUT="${TIMEOUT:-90s}"
+USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-true}"
```
Comment (ksubrmnn) [inline demo/run-osm-demo.sh:40]:
```suggestion
USE_PRIVATE_REGISTRY="${USE_PRIVATE_REGISTRY:-false}"
```

File: demo/run-osm-demo.sh [hunk 12]
```diff
@@ -86,6 +87,9 @@ fi
 
 if [ "$PUBLISH_IMAGES" = true ]; then
     make docker-push
+fi 
+
+if [ "$USE_PRIVATE_REGISTRY" = true ]; then
     ./scripts/create-container-registry-creds.sh "$K8S_NAMESPACE"
```
Comment (ksubrmnn) [inline demo/run-osm-demo.sh:93]: You need to also do the same thing that you do here in the configure-app-namespaces script.

Comment (nshankar13) [reply demo/run-osm-demo.sh:93]: I address that in this PR: https://github.com/openservicemesh/osm/pull/3876

Comment (nshankar13) [reply demo/run-osm-demo.sh:93]: ^ realized I need to update the default value of USE_PRIVATE_REGISTRY there to false as well

File: .env.example [hunk 13]
```diff
@@ -170,6 +170,10 @@ export BOOKWAREHOUSE_NAMESPACE=bookwarehouse
 # Default: 90s
 # export TIMEOUT=90s
 
+### optional: Whether the CTR_REGISTRY is a private container registry
+# Default: false
+#export USE_PRIVATE_REGISTRY=false
```
Comment (ksubrmnn) [inline .env.example:175]: @nojnhuh do you think this should be true or false by default? I think we should match whatever kind needs. Do we need the ctr_registry creds for kind?

Comment (nojnhuh) [reply .env.example:175]: The demo itself doesn't manage kind clusters like the e2e tests do, the demo always assumes you have _some_ cluster up and running beforehand. I usually do `make kind-up` which also ensures a local image registry is running and makes it available to the kind cluster. That registry doesn't require credentials and since it's the default in .env.example I think false for this by default makes the most sense.

I admit I haven't been following this PR very closely, but scripts/create-container-registry-creds.sh already checks whether or not the image pull secret needs to be created based on if `$CTR_REGISTRY_PASSWORD` is defined: https://github.com/openservicemesh/osm/blob/de4c5b9eb941aa09147005be0224d6c686908022/scripts/create-container-registry-creds.sh#L16

Does `$USE_PRIVATE_REGISTRY` mean that same thing? If so, I don't have a preference which we use to check if the secrets are needed, but it would be worth making that consistent.

Comment (nshankar13) [reply .env.example:175]: We thought that using a variable like USE_PRIVATE_REGISTRY would be a little more explicit and intuitive than checking to see if the CTR_REGISTRY_PASSWORD is null. But yeah it essentially does the same thing. 

I also opened this PR separately to use the USE_PRIVATE_REGISTRY variable in other scripts as well: https://github.com/openservicemesh/osm/pull/3876

File: demo/deploy-bookstore.sh [hunk 14]
```diff
@@ -39,8 +40,10 @@ metadata:
 EOF
 
 if [ "$DEPLOY_ON_OPENSHIFT" = true ] ; then
-    oc adm policy add-scc-to-user privileged -z "$SVC" -n "$BOOKSTORE_NAMESPACE"
-    oc secrets link "$SVC" "$CTR_REGISTRY_CREDS_NAME" --for=pull -n "$BOOKSTORE_NAMESPACE"
+    oc adm policy add-scc-to-user privileged -z bookstore -n "$BOOKSTORE_NAMESPACE"
```
Comment (ksubrmnn) [inline demo/deploy-bookstore.sh:43]: These should still be $SVC instead of bookstore since this script is used to configure the  bookstore-v1 and bookstore-v2 service accounts

Comment (ksubrmnn) [reply demo/deploy-bookstore.sh:43]:
```suggestion
    oc adm policy add-scc-to-user privileged -z "$SVC" -n "$BOOKSTORE_NAMESPACE"
```

**Review by ksubrmnn — APPROVED** [review]

File: demo/deploy-bookbuyer.sh [hunk 15]
```diff
@@ -24,7 +25,9 @@ EOF
 
 if [ "$DEPLOY_ON_OPENSHIFT" = true ] ; then
     oc adm policy add-scc-to-user privileged -z bookbuyer -n "$BOOKBUYER_NAMESPACE"
-    oc secrets link bookbuyer "$CTR_REGISTRY_CREDS_NAME" --for=pull -n "$BOOKBUYER_NAMESPACE"
+    if [ "$USE_PRIVATE_REGISTRY" = true ]; then
+        oc secrets link bookbuyer "$CTR_REGISTRY_CREDS_NAME" --for=pull -n "$BOOKBUYER_NAMESPACE"
```
Comment (draychev) [inline demo/deploy-bookbuyer.sh:29]: 👍🏻

**Review by draychev — APPROVED** [review]

File: .env.example (hunk 13 above)
Comment (snehachhabria) [reply .env.example:175]: could someone please explain the intention of this change ? If openshift specifically doesn't need a private registry then this can be explicitly be set to false for that particular scenario and not for the demo in general.

If someone is using a private registry for development purpose, which is likely the case ( excluding kind) this will cause issues.

Comment (snehachhabria) [reply .env.example:175]: Further while making such changes please ensure that it is tested not only on kind but on other cluster types as well

Comment (snehachhabria) [reply .env.example:175]: If scripts/create-container-registry-creds.sh already checks whether or not the image pull secret needs to be created based on if $CTR_REGISTRY_PASSWORD is defined, please define USE_PRIVATE_REGISTRY based on that (by adding custom logic) and do not define another default value, expecting users to set another variable in their .env.

Comment (nshankar13) [reply .env.example:175]: Discussed offline with Sneha, we are thinking it might be better to set the default value of USE_PRIVATE_REGISTRY to true.
